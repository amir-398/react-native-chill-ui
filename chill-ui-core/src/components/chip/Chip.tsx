import { cn } from '@utils';

import { classNamePropsHandler } from '@/utils/hybrid/classNamePropsHandler';
import { classNameHandler, styleHandler } from '@/utils/hybrid/propsHandlers';
import colorVariantPropsHandler from '@/utils/hybrid/colorVariantPropsHandler';

import type { ChipProps } from '../../types/chip.types';

import Icon from '../icon';
import { Box } from '../box';
import { String } from '../string';
import createChipStyles from './utils/createStyles';
import { isNativeWindInstalled } from '../../utils/hybrid/nativewindDetector';
import chipVariants, { chipPositionVariants, chipTextVariants } from './Chip.variants';

/**
 * Chip component displays compact elements that represent an input, attribute, or action.
 * Supports both text and icon content with customizable colors, sizes, border radius variants,
 * and style variants (contained/outlined). Automatically detects NativeWind availability
 * and falls back to StyleSheet if needed.
 *
 * @example
 * ```tsx
 * <Chip variant="contained" size="md" rounded="full">
 *   New
 * </Chip>
 *
 * <Chip
 *   variant="outlined"
 *   color="#3B82F6"
 *   size="lg"
 *   rounded="full"
 * >
 *   Custom
 * </Chip>
 *
 * <Chip
 *   variant="contained"
 *   iconActions={[{ iconName: 'star-solid', position: 'left' }]}
 *   size="lg"
 *   rounded="full"
 * />
 * ```
 *
 * @param color - Custom background color for the chip
 * @param children - Content to display in the chip
 * @param className - Custom CSS classes (only used with NativeWind)
 * @param iconActions - Icon actions configuration
 * @param rounded - Border radius variant (default: 'full')
 * @param size - Size variant for the chip (default: 'md')
 * @param position - Position variant for the chip (default: 'left')
 * @param stringProps - Props for the string/text component
 * @param title - Title to display in the chip (priority over children)
 * @param variant - Style variant: 'contained' (filled) or 'outlined' (bordered) - default: 'contained'
 *
 * @see {@link https://mui.com/material-ui/react-chip/ MUI Chip Documentation}
 */
export default function Chip(props: ChipProps) {
  classNamePropsHandler(props, 'Chip');
  colorVariantPropsHandler(props, 'Chip');
  const {
    children,
    className,
    color,
    colorVariant = 'primary',
    iconActions = [],
    position = 'left',
    rounded = 'full',
    size = 'md',
    stringProps,
    style,
    title,
    variant = 'contained',
    ...rest
  } = props;

  const isNativeWind = isNativeWindInstalled();

  const isChipHasIcon: boolean =
    iconActions.length > 0 &&
    iconActions.some(action => action.customIcon || action.iconName) &&
    iconActions.some(
      action => action.position === 'left' || action.position === 'right' || action.position === 'center',
    );

  const chipStyles = !isNativeWind ? createChipStyles(variant, rounded, size, color, isChipHasIcon) : null;

  /**
   * Renders icons based on their position
   */
  const renderIcons = (iconPosition: 'left' | 'right' | 'center') => {
    const icons = iconActions.filter(action => action.position === iconPosition);

    const getIconClassName = (pos: 'left' | 'right' | 'center') => {
      if (pos === 'left') return 'mr-1.5';
      if (pos === 'right') return 'ml-1.5';
      return 'mx-1.5';
    };

    const getIconStyle = (pos: 'left' | 'right' | 'center') => {
      if (pos === 'left') return { marginRight: 6 };
      if (pos === 'right') return { marginLeft: 6 };
      return { marginLeft: 3, marginRight: 3 };
    };

    return icons.map((action, index) => {
      if (action.customIcon) {
        return (
          <Box
            key={`${iconPosition}-${index}`}
            {...classNameHandler(getIconClassName(iconPosition))}
            {...styleHandler({ defaultStyle: getIconStyle(iconPosition) })}
          >
            {action.customIcon}
          </Box>
        );
      }

      if (action.iconName) {
        return (
          <Icon
            key={`${iconPosition}-${index}`}
            name={action.iconName as any}
            size={(action.iconSize as any) ?? (size as any) ?? 'sm'}
            color={action.iconColor}
            {...classNameHandler(getIconClassName(iconPosition))}
            {...styleHandler({ defaultStyle: getIconStyle(iconPosition) })}
          />
        );
      }

      return null;
    });
  };

  /**
   * Renders the main content of the chip
   */
  const renderContent = () => {
    const getTextClassName = () => {
      if (variant === 'outlined' && color) {
        return '';
      }
      return cn(chipTextVariants({ colorVariant, variant }));
    };

    const getTextStyle = () => {
      if (variant === 'outlined' && color) {
        return { ...chipStyles?.text, color };
      }
      return chipStyles?.text;
    };

    if ((children && typeof children === 'string') || title) {
      return (
        <String
          size={size as any}
          {...classNameHandler(getTextClassName())}
          {...styleHandler({ defaultStyle: getTextStyle() })}
          {...stringProps}
        >
          {title ?? children}
        </String>
      );
    }
    return children;
  };

  return (
    <Box
      {...classNameHandler(
        cn(
          'self-center px-3 py-2',
          chipVariants({ colorVariant, rounded, size: size as any, variant }),
          chipPositionVariants({ position }),
          {
            'flex-row items-center': isChipHasIcon,
          },
          className,
        ),
      )}
      {...styleHandler({
        defaultStyle: [chipStyles?.chip, { ...(color && { backgroundColor: color, borderColor: color }) }, style],
      })}
      {...rest}
    >
      {renderIcons('left')}
      {renderContent()}
      {renderIcons('center')}
      {renderIcons('right')}
    </Box>
  );
}
