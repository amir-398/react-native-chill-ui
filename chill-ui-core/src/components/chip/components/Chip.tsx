import type { ChipPropsTw, IconProps } from '@types';

import { Box } from '@components/box';
import { Icon } from '@components/icon';
import { String } from '@components/string';
import { Pressable, TouchableOpacity } from 'react-native';
import { ScalePressable } from '@components/scalePressable';
import { RipplePressable } from '@components/ripplePressable';
import { classNameHandler, styleHandler, classNamePropsHandler, cn, colorVariantPropsHandler } from '@utils';

import { chipDefaultProps } from '../utils/defaultProps';
import { chipSv, styles } from '../styles/Chip.ss.styles';
import { chipTv, chipTextTv, twStyles } from '../styles/Chip.tw.styles';

/**
 * Chip component displays compact elements that represent an input, attribute, or action.
 * Supports both text and icon content with customizable colors, sizes, and style variants (contained/outlined).
 * Automatically detects NativeWind availability and falls back to StyleSheet if needed.
 *
 * @example
 * ```tsx
 * <Chip
 *   variant="outlined"
 *   colorVariant="primary"
 *   size="md"
 *   onPress={handlePress}
 *   leftIconAction={{ name: 'star-solid' }}
 * >
 *   New Feature
 * </Chip>
 * ```
 *
 * @param as - Type of touchable component to use when onPress is provided (default: 'touchable-opacity')
 * @param children - Content to display in the chip
 * @param className - (only NativeWind) Additional CSS classes for the chip container
 * @param color - Custom background color for the chip
 * @param colorVariant - (only NativeWind) Color variant: 'primary' | 'secondary' | 'accent' | 'danger' | 'warning' | 'info' | 'success' | 'error' | 'dark' | 'light' | 'muted' | 'neutral' | 'disabled' | 'inverted' | 'white' (default: 'primary')
 * @param leftIconAction - Left icon configuration (name, size, color, customIcon)
 * @param onPress - Press callback function for the chip
 * @param position - Position variant for the chip: 'left' | 'center' | 'right' (default: 'left')
 * @param rightIconAction - Right icon configuration (name, size, color, customIcon)
 * @param size - Size variant: 'xs' | 'sm' | 'md' | 'lg' | 'xl' | '2xl' | '2xs' (default: 'xs')
 * @param stringProps - Props for the String component
 * @param style - Style object for the chip container
 * @param title - Title to display in the chip (priority over children)
 * @param variant - Style variant: 'contained' | 'outlined' (default: 'contained')
 */
export default function Chip(props: ChipPropsTw) {
  classNamePropsHandler(props, 'Chip');
  colorVariantPropsHandler(props, 'Chip');
  const {
    as = chipDefaultProps.as,
    children,
    className,
    color,
    colorVariant = chipDefaultProps.colorVariant,
    leftIconAction,
    onPress,
    position = chipDefaultProps.position,
    rightIconAction,
    size = chipDefaultProps.size,
    stringProps,
    style,
    title,
    variant = chipDefaultProps.variant,
    ...rest
  } = props;

  const hasLeftIcon = Boolean(leftIconAction && (leftIconAction.customIcon || leftIconAction.name));
  const hasRightIcon = Boolean(rightIconAction && (rightIconAction.customIcon || rightIconAction.name));
  const isChipHasIcon = hasLeftIcon || hasRightIcon;

  /**
   * Renders a single icon based on configuration
   */
  const renderIcon = (
    iconAction: ChipPropsTw['leftIconAction'] | ChipPropsTw['rightIconAction'],
    iconPosition: 'left' | 'right',
  ) => {
    if (!iconAction) return null;

    const iconClassName = iconPosition === 'left' ? twStyles.iconLeft : twStyles.iconRight;
    const iconStyle = iconPosition === 'left' ? styles.iconLeft : styles.iconRight;

    if (iconAction.customIcon) {
      return (
        <Box
          {...classNameHandler(cn(iconClassName, twStyles.pointerEventsNone))}
          {...styleHandler({ defaultStyle: [iconStyle, styles.pointerEventsNone] })}
        >
          {iconAction.customIcon}
        </Box>
      );
    }

    if (iconAction.name) {
      return (
        <Box
          {...classNameHandler(cn(iconClassName, twStyles.pointerEventsNone))}
          {...styleHandler({ defaultStyle: [iconStyle, styles.pointerEventsNone] })}
        >
          <Icon
            name={iconAction.name}
            size={iconAction.size ?? (size as IconProps['size']) ?? 'sm'}
            color={iconAction.color}
          />
        </Box>
      );
    }

    return null;
  };

  /**
   * Renders the main content of the chip
   */
  const renderContent = () => {
    if ((children && typeof children === 'string') || title) {
      return (
        <String
          size={size}
          {...stringProps}
          {...classNameHandler(cn(chipTextTv({ colorVariant, variant }), stringProps?.className))}
          {...styleHandler({
            style: [{ ...(variant === 'outlined' && color && { color }) }, stringProps?.style],
          })}
        >
          {title ?? children}
        </String>
      );
    }
    return children;
  };

  const chipContent = (
    <>
      {renderIcon(leftIconAction, 'left')}
      {renderContent()}
      {renderIcon(rightIconAction, 'right')}
    </>
  );

  const commonProps = {
    ...classNameHandler(
      cn(
        twStyles.chip,
        chipTv({ colorVariant, position, variant }),
        {
          [twStyles.chipWithIcons]: isChipHasIcon,
        },
        className,
      ),
    ),
    ...styleHandler({
      defaultStyle: [styles.chip, chipSv({ position, variant }), isChipHasIcon && styles.chipWithIcons],
      style: [
        { ...(color && { backgroundColor: variant === 'outlined' ? 'transparent' : color, borderColor: color }) },
        style,
      ],
    }),
    ...rest,
  };

  // Si pas de onPress, on retourne juste une Box
  if (!onPress) {
    return (
      <Box>
        <Box {...commonProps}>{chipContent}</Box>
      </Box>
    );
  }

  // Si onPress est fourni, on wrappe dans le composant touchable appropri√©
  switch (as) {
    case 'ripple-pressable':
      return (
        <Box>
          <RipplePressable {...commonProps} onPress={onPress}>
            {chipContent}
          </RipplePressable>
        </Box>
      );

    case 'pressable':
      return (
        <Box>
          <Pressable {...commonProps} onPress={onPress}>
            {chipContent}
          </Pressable>
        </Box>
      );

    case 'scale-pressable':
      return (
        <Box>
          <ScalePressable {...commonProps} onPress={onPress}>
            {chipContent}
          </ScalePressable>
        </Box>
      );

    case 'touchable-opacity':
    default: {
      return (
        <Box>
          <TouchableOpacity {...commonProps} onPress={onPress} activeOpacity={0.7}>
            {chipContent}
          </TouchableOpacity>
        </Box>
      );
    }
  }
}
