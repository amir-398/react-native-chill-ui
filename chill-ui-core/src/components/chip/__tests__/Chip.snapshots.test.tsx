import { render } from '@testing-library/react-native';

import Chip from '../components/Chip';

// Mocks
jest.mock('../../../utils', () => ({
  classNameHandler: jest.fn(() => ({})),
  classNamePropsHandler: jest.fn(),
  cn: jest.fn((...args) => args.filter(Boolean).join(' ')),
  colorVariantPropsHandler: jest.fn(),
  isString: jest.fn(value => typeof value === 'string'),
  styleHandler: jest.fn(() => ({})),
}));

jest.mock('../../../components/box', () => ({
  Box: ({ children, ...props }: any) => <div {...props}>{children}</div>,
}));

jest.mock('../../../components/icon', () => ({
  Icon: ({ name, ...props }: any) => <div data-testid={`icon-${name}`} {...props} />,
}));

jest.mock('../../../components/string', () => ({
  String: ({ children, ...props }: any) => <div {...props}>{children}</div>,
}));

jest.mock('../../../components/ripplePressable', () => ({
  RipplePressable: ({ children, ...props }: any) => <div {...props}>{children}</div>,
}));

jest.mock('../../../components/scalePressable', () => ({
  ScalePressable: ({ children, ...props }: any) => <div {...props}>{children}</div>,
}));

jest.mock('../styles/Chip.ss.styles', () => ({
  chipSv: jest.fn(() => ({})),
  chipTextSv: jest.fn(() => ({})),
  styles: { chip: {}, chipWithIcons: {}, iconLeft: {}, iconRight: {}, pointerEventsNone: {} },
}));

jest.mock('../styles/Chip.tw.styles', () => ({
  chipTextTv: jest.fn(() => ''),
  chipTv: jest.fn(() => ''),
  twStyles: { chip: '', chipWithIcons: '', iconLeft: '', iconRight: '', pointerEventsNone: '' },
}));

describe('Chip Snapshots', () => {
  it('should match snapshot for basic variants', () => {
    const tree = render(
      <div>
        <Chip variant="contained">Contained</Chip>
        <Chip variant="outlined">Outlined</Chip>
      </div>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot for sizes', () => {
    const tree = render(
      <div>
        <Chip size="xs">XS</Chip>
        <Chip size="md">MD</Chip>
        <Chip size="lg">LG</Chip>
      </div>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot for icons', () => {
    const tree = render(
      <div>
        <Chip leftIconAction={{ name: 'star-solid' }}>Left Icon</Chip>
        <Chip rightIconAction={{ name: 'close-solid' }}>Right Icon</Chip>
      </div>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot for positions', () => {
    const tree = render(
      <div>
        <Chip position="left">Left</Chip>
        <Chip position="center">Center</Chip>
        <Chip position="right">Right</Chip>
      </div>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
