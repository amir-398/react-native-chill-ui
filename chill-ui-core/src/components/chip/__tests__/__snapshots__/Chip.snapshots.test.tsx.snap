// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chip Snapshots should match snapshot for basic variants 1`] = `
<div>
  <div>
    <div>
      <div
        size="xs"
      >
        Contained
      </div>
    </div>
  </div>
  <div>
    <div>
      <div
        size="xs"
      >
        Outlined
      </div>
    </div>
  </div>
</div>
`;

exports[`Chip Snapshots should match snapshot for icons 1`] = `
<div>
  <div>
    <div>
      <div>
        <div
          data-testid="icon-star-solid"
          size="xs"
        />
      </div>
      <div
        size="xs"
      >
        Left Icon
      </div>
    </div>
  </div>
  <div>
    <div>
      <div
        size="xs"
      >
        Right Icon
      </div>
      <div>
        <div
          data-testid="icon-close-solid"
          size="xs"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Chip Snapshots should match snapshot for positions 1`] = `
<div>
  <div>
    <div>
      <div
        size="xs"
      >
        Left
      </div>
    </div>
  </div>
  <div>
    <div>
      <div
        size="xs"
      >
        Center
      </div>
    </div>
  </div>
  <div>
    <div>
      <div
        size="xs"
      >
        Right
      </div>
    </div>
  </div>
</div>
`;

exports[`Chip Snapshots should match snapshot for sizes 1`] = `
<div>
  <div>
    <div>
      <div
        size="xs"
      >
        XS
      </div>
    </div>
  </div>
  <div>
    <div>
      <div
        size="md"
      >
        MD
      </div>
    </div>
  </div>
  <div>
    <div>
      <div
        size="lg"
      >
        LG
      </div>
    </div>
  </div>
</div>
`;
